### PR TITLE
Improve kw_only_dataclass

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

* Checks the number of positional arguments on `__init__` to prevent assigning `kw_only` arguments.
* Fixes existing tests that leverage positional leaking into kw_only.
* Fixes mypy erros

### CODE REVIEW

**Comment by cgarciae** [comment]
@superbobry

File: flax/linen/kw_only_dataclasses.py [hunk 1]
```diff
@@ -122,7 +130,8 @@ def _process_class(cls, extra_fields=None, **kwargs):
   base_dataclass_fields = {}  # dict[cls, cls.__dataclass_fields__.copy()]
 
   # The keyword only fields from `cls` or any of its base classes.
-  kw_only_fields = {}  # dict[field_name, tuple[annotation, default]]
+  # dict[field_name, tuple[annotation, default]]
```
Comment (superbobry) [inline flax/linen/kw_only_dataclasses.py:133]: This type in a comment feels a bit odd, now that `kw_only_fields` is annotated. Feel free to keep as is, but alternatively you could

* merge it into the human readable line above: "A mapping of keyword-only fields ... to a pair of ...";
* introduce type aliases, e.g. `Annotation = Any` etc;
* just drop it entirely :)

**Review by superbobry — APPROVED** [review]
Nice!

File: flax/linen/kw_only_dataclasses.py [hunk 2]
```diff
@@ -111,7 +119,7 @@ def wrap(cls):
   return wrap if cls is None else wrap(cls)
 
 
-def _process_class(cls, extra_fields=None, **kwargs):
+def _process_class(cls: Type[M], extra_fields=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/kw_only_dataclasses.py:122]: right!

File: flax/linen/kw_only_dataclasses.py [hunk 3]
```diff
@@ -190,12 +201,36 @@ def _process_class(cls, extra_fields=None, **kwargs):
...
   # Restore the base classes' __dataclass_fields__.
-  for cls, dataclass_fields in base_dataclass_fields.items():
-    cls.__dataclass_fields__ = dataclass_fields
+  for _cls, fields in base_dataclass_fields.items():
+    _cls.__dataclass_fields__ = fields
+
+  if create_init:
+    dataclass_init = transformed_cls.__init__
+    _fields = dataclasses.fields(transformed_cls)
+    expected_num_args = len(
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/kw_only_dataclasses.py:216]: I wish there was a function in `itertools` that return the length of an iterator by consuming it.
I'll use your solution and add a comment :)

File: flax/linen/kw_only_dataclasses.py (hunk 1 above)
Comment (cgarciae) [reply flax/linen/kw_only_dataclasses.py:133]: Ok, did the future a favor and added type aliases.

File: flax/linen/kw_only_dataclasses.py [hunk 4]
```diff
@@ -138,7 +147,9 @@ def _process_class(cls, extra_fields=None, **kwargs):
         )
       default = getattr(cls, name)
       if isinstance(default, dataclasses.Field):
-        default.metadata = {**default.metadata, **{KW_ONLY: True}}
+        default.metadata = MappingProxyType(
+            {**default.metadata, **{KW_ONLY: True}}
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/kw_only_dataclasses.py:151]: Not sure what I was thinking at the moment here 😅

File: flax/linen/kw_only_dataclasses.py [hunk 5]
```diff
@@ -88,6 +95,7 @@ def field(*, metadata=None, kw_only=dataclasses.MISSING, **kwargs):
   return dataclasses.field(metadata=metadata, **kwargs)
 
 
+@dataclass_transform(field_specifiers=(field,))  # type: ignore[literal-required]
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/kw_only_dataclasses.py:98]: No sure what is the deal with `mypy` here. Especially that the error is `literal-required`.
We been just been doing this on every `dataclass_transform` in the code base.

**Comment by cgarciae** [comment]
Thanks @superbobry for the detailed review! I've resolved almost all the comments except for the early return one which I'll leave it as is.

**Pull request merged**